entity-service-client/src/main/java/org/hypertrace/entity/data/service/client/EdsCacheClient.java

```diff
@@ -10,47 +10,83 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class EdsCacheClient implements EdsClient {
 
   private static final Logger LOG = LoggerFactory.getLogger(EdsCacheClient.class);
-  private final LoadingCache<EdsCacheKey, Optional<EnrichedEntity>> enrichedEntityCache;
-  private final LoadingCache<EdsCacheKey, Optional<Entity>> entityCache;
+  private LoadingCache<EdsCacheKey, EnrichedEntity> enrichedEntityCache;
+  private LoadingCache<EdsCacheKey, Entity> entityCache;
+  private LoadingCache<EdsTypeAndIdAttributesCacheKey, String> entityIdsCache;
 
   private final EntityDataServiceClient client;
 
   EdsCacheClient(EntityServiceClientConfig entityServiceClientConfig) {
-    this(new EntityDataServiceClient(entityServiceClientConfig));
+    this(new EntityDataServiceClient(entityServiceClientConfig),
+        entityServiceClientConfig.getCacheConfig());
   }
 
-  public EdsCacheClient(EntityDataServiceClient client) {
+  public EdsCacheClient(EntityDataServiceClient client,
+      EntityServiceClientCacheConfig cacheConfig) {
     this.client = client;
+    initCache(cacheConfig);
+  }
+
+  private void initCache(EntityServiceClientCacheConfig cacheConfig) {
     this.enrichedEntityCache =
-        CacheBuilder.newBuilder().expireAfterWrite(5, TimeUnit.MINUTES).maximumSize(1000)
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(cacheConfig.getEnrichedEntityCacheExpiryMs(), TimeUnit.MILLISECONDS)
+            .maximumSize(cacheConfig.getEnrichedEntityMaxCacheSize())
             .build(new CacheLoader<>() {
-              public Optional<EnrichedEntity> load(@Nonnull EdsCacheKey key) {
-                return Optional
-                    .ofNullable(client.getEnrichedEntityById(key.tenantId, key.entityId));
+              public EnrichedEntity load(@Nonnull EdsCacheKey key) throws Exception {
+                EnrichedEntity enrichedEntity = client
+                    .getEnrichedEntityById(key.tenantId, key.entityId);
+                if (enrichedEntity == null) {
+                  throw new NotFoundException("Enriched entity not found");
+                }
+                return enrichedEntity;
               }
             });
 
     this.entityCache =
-        CacheBuilder.newBuilder().expireAfterWrite(5, TimeUnit.MINUTES).maximumSize(1000)
+        CacheBuilder.newBuilder()
+            .expireAfterWrite(cacheConfig.getEntityCacheExpiryMs(), TimeUnit.MILLISECONDS)
+            .maximumSize(cacheConfig.getEntityMaxCacheSize())
             .build(new CacheLoader<>() {
-              public Optional<Entity> load(@Nonnull EdsCacheKey key) {
-                return Optional.ofNullable(client.getById(key.tenantId, key.entityId));
+              public Entity load(@Nonnull EdsCacheKey key) throws Exception {
+                Entity entity = client.getById(key.tenantId, key.entityId);
+                if (entity == null) {
+                  throw new NotFoundException("Entity not found");
```

Comment:
when can this happen ?

Reply:
It's possible that gateway service passes an invalid entityID from public Graphql query or Explorer API.

Reply:
got it